### PR TITLE
Introduce GQLDocument

### PR DESCRIPTION
**Pull request by martinbonnin**
Replace SDL parsing with `GQLDocument`, a simple GraphQL AST 

**Fixes the below commits from main:**
60d7e1c662f92cc111c45dcfcf9a95e8505c69b7 optional mutations/subscriptions
f3dca0383980359ff453749a74748f140daf8eb3 type extensions

This PR is dependant on https://github.com/apollographql/apollo-android/pull/2724

### CODE REVIEW

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/antlr2ast.kt [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+import org.antlr.v4.runtime.Token
+import org.antlr.v4.runtime.tree.TerminalNode
+
+
+private fun SourceLocation(token: Token) = SourceLocation(
+    line = token.line,
+    position = token.charPositionInLine
+)
+
+
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/antlr2ast.kt:15]: nit: extra blank lines ^^^^

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/ast2introspection.kt [hunk 2]
```diff
@@ -0,0 +1,250 @@
+import com.apollographql.apollo.compiler.parser.error.ParseException
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/ast2introspection.kt:1]: Should we call this file `IntrospectionSchemaBuilder.kt`

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt [hunk 3]
```diff
@@ -0,0 +1,128 @@
+package com.apollographql.apollo.compiler.parser.graphql.ast
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt:1]: nit: is there a way to better name this file, like `documentParser`?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+fun GQLDocument.withoutBuiltinTypes(): GQLDocument {
+  return copy(
+      definitions = definitions.filter {
+        when ((it as? GQLNamed)?.name) {
+          "Int", "Float", "String", "ID", "Boolean", "__Schema",
+          "__Type", "__Field", "__InputValue", "__EnumValue", "__TypeKind",
+          "__Directive", "__DirectiveLocation" -> false
+          else -> true
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt:36]: nit: can it be shorter `.filterNot { (it as? GQLNamed)?.name in listOf(....) }`

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        }
+      }
+  )
+}
+
+/**
+ * Plain parsing, without validation or adding the builtin types
+ */
+private fun GQLDocument.Companion.parseInternal(inputStream: InputStream, filePath: String = "(source)"): GQLDocument {
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt:45]: `= "(source)"` ?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/introspection2ast.kt [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        name = name,
+        directives = emptyList()
+    )
+  }
+}
+
+fun IntrospectionSchema.toDocument(): GQLDocument = GQLDocumentBuilder(this).toGQLDocument()
+
+
+
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/introspection2ast.kt:216]: nit: extra lines

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/introspection2ast.kt [hunk 7]
```diff
@@ -0,0 +1,216 @@
+package com.apollographql.apollo.compiler.parser.graphql.ast
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/introspection2ast.kt:1]: nit: do you think should this file called `GQLDocumentBuilder `

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/introspection2ast.kt [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    return GQLInputObjectTypeDefinition(
+        description = description,
+        name = name,
+        inputFields = inputFields.map { it.toGQLInputValueDefinition() },
+        directives = emptyList()
+    )
+  }
+
+  private fun IntrospectionSchema.InputField.toGQLInputValueDefinition(): GQLInputValueDefinition
+  {
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/introspection2ast.kt:194]: nit: codestyle

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/schema_validation.kt [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+  validateNotExecutable()
+  validateUniqueSchemaDefinition()
+  validateTypeNames()
+  validateDirectiveNames()
+  validateInterfaces()
+  validateObjects()
+  return this
+}
+
+fun GQLDocument.mergeTypeExtensions(): GQLDocument {
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/schema_validation.kt:15]: is it a right place for this function? Looks like all others related to validation this one is merging definitions

**Review by sav007 — APPROVED** [review]

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/antlr2ast.kt (hunk 1 above)
Comment (martinbonnin) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/antlr2ast.kt:15]: This should be removed in #2756

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/ast2introspection.kt (hunk 2 above)
Comment (martinbonnin) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/ast2introspection.kt:1]: I didn't pay super attention to the names of the `internal`/`private` types but that definitely requires some love. A pattern that I see emerging is classes that don't serve any other purpose than holding a `schema` and `fragments`. I started naming these `XXXScope`, `InputValueValidationScope`, `MergeTypeExtensionsScope`, etc.. No strong feeling there so I'm happy to change to `XXXBuilder` or something else.

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt (hunk 3 above)
Comment (martinbonnin) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt:1]: I renamed to `GraphQLParser` in #2756

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt (hunk 4 above)
Comment (martinbonnin) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt:36]: Rewrote it in #2756 : https://github.com/apollographql/apollo-android/blob/ec9190dd77d6421cd7878dc3bc86ab83e7aa6f9f/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/gql/gqltypedefinition.kt#L25

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt (hunk 5 above)
Comment (martinbonnin) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/graphql/ast/document.kt:45]: Yea, that's weird, if we parse from a String for an example, we cannot display the nice "preview". I guess I should just make filePath optional

**Comment by martinbonnin** [comment]
Merged as part of https://github.com/apollographql/apollo-android/pulls?q=is%3Apr+is%3Aclosed+gqldocument

**Pull request closed without merging**